                                             Charakterystyka zmiennych:
                                             
                                            Characteristics of variables:
                                             
                                             

vis - widoczność (km)
      visibility

year - rok, z którego pochodzą pomiary
       the year from which the neasurements come from

temp - temperatura w stopniach Celsjusza
       temperature in degrees Celsius

hum - wilgotność
      humidity

windsp - prędkość wiatru (km/h)
         wind speed

pres - ciśnienie (millibar)
       pressure

rain - deszcz

snow - śnieg

mosclou - przeważające zachmurzenie
          mostlu cloudy

parclou - częściowe zachmurzenie
          partly cloudy

overcast - całkowite zachmurzenie
           completely cloudy

clear - bezchmurnie
        cloudless

foggy - mgła

 
 Dane pochodzą z lat 2006 - 2016. Dotyczą warunków atmosferycznych w mieście Segedyn na Węgrzech.
 The data comes from years 2006 - 2016. They concern weather conditions in the city Szeged in Hungary. 
 
 
 Cel projektu: próba stworzenia narzędzia potrafiącego prognozować warunki widoczności. 
 Project goal: an attempt to create a tool that can forecast visibility conditions.
 

In [1]:
import numpy as np
import pandas as pd

In [2]:
dfMod = pd.read_csv('Dane/dfMod.csv')
dfMod.head()

year  temp   hum  windsp    vis     pres  rain  snow  mosclou  parclou   
0  2006  9.47  0.89   14.12  15.83  1015.13     1     0        0        1  \
1  2006  9.36  0.86   14.26  15.83  1015.63     1     0        0        1   
2  2006  9.38  0.89    3.93  14.96  1015.94     1     0        1        0   
3  2006  8.29  0.83   14.10  15.83  1016.41     1     0        0        1   
4  2006  8.76  0.83   11.04  15.83  1016.51     1     0        1        0   

   overcast  clear  foggy  
0         0      0      0  
1         0      0      0  
2         0      0      0  
3         0      0      0  
4         0      0      0

                                        Regresja liniowa jednej zmiennej:
                                        
                                        Linear regression of one variable:

In [3]:
import sklearn
from sklearn.linear_model import LinearRegression 

In [4]:
Xl = dfMod[['temp']]
yl = dfMod['vis']

In [5]:
Xl.values[:5]

array([[9.47],
       [9.36],
       [9.38],
       [8.29],
       [8.76]])

In [6]:
yl.values[:5]

array([15.83, 15.83, 14.96, 15.83, 15.83])

In [7]:
model1 = LinearRegression()
model1.fit(Xl, yl)

LinearRegression()

In [8]:
model1.intercept_

8.312394156362561

In [9]:
model1.coef_

array([0.17168457])

In [10]:
predict_data1 = np.array([[15.5], [16.7]])

In [11]:
predict_data1

array([[15.5],
       [16.7]])

In [12]:
model1.predict(predict_data1)

C:\Users\Kamil - IT\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([10.97350505, 11.17952654])

                                              Metryki oceny regresji:                                             

 Ważniejsze metryki oceny regresji:
    
  R2 (współczynnik determinacji) - jest opisową miarą siły liniowego związku między zmiennymi, tj. miarą dopasowania linii
                                  regresji do danych. Przyjmuje wartości z przedziału [0, 1] i wskazuje jaka część zmiennej 'y'
                                  jest wyjaśniana przez wybrany model. 
            
  MAE (Mean Absolute Error) - mierzy średnią różnicę pomiędzy wartościami przewidzianymi a rzeczywistymi. Należy pamiętać,
                              że poprzez wartość absolutną tracimy informację o kierunku (wartość wyższa bądź niższa od
                              oczekiwanej). Nadto, również wartości odstające mogą w znaczący sposób wpływać na wynik.
        
  MAPE (Mean Absolute Percentage Error) - mierzy procentową wartość błędu predykcji. 
    
  MSE (Mean Squared Error) - czyli średni błąd kwadratowy, mierzy średnią kwadratową różnicę między wartościami oszacowanymi
                             a rzeczywistymi. Zawsze przyjmuje wartość dodatnią, która maleje w miarę zbliżania się błędu do 0.
        
  RMSE (Root Mean Squared Error) - pierwiatek kwadratowy z MSE.
                                     

                                           Regression evaluation metrics:
                                           
 Important regression evaluation metrics:
 
  R2 (coefficient of determination) - is a descriptive measure of the strength of the linear relationship between variables,
                                      i.e., a measure of line fit regression to data. It accepts values in the range [0, 1]
                                      and indicates what part of the variable 'y' is explained by the chosen model.
                                      
  MAE - measures the average difference between prediccted and actual values. Please remember that through the absolute value
        we lose information about the direction (value higher or lower than expected). Moreover, outlier values may affect
        significantly the result.
        
  MAPE - measures the percentage of prediction error.
  
  MSE - measures the mean squared difference between estimated values and real ones. It always takes a positive value, which
        decreases as the error approaches 0.
        
  RMSE - square root of MSE. 
                                           

In [13]:
from sklearn.metrics import r2_score

In [14]:
yl_predict = model1.predict(Xl)

In [15]:
print("R2:", r2_score(yl, yl_predict))

R2: 0.15496386646453242


In [16]:
from sklearn.metrics import mean_absolute_error

In [17]:
yl_pred = model1.predict(Xl)

In [18]:
print("MAE:", mean_absolute_error(yl, yl_pred))

MAE: 3.1175224185835857


In [101]:
from sklearn.model_selection import train_test_split

In [102]:
Xl_train, Xl_test, yl_train, yl_test = train_test_split(Xl, yl, test_size=0.2, random_state=42)

In [103]:
model1a = LinearRegression()
model1a.fit(Xl_train, yl_train)

LinearRegression()

In [104]:
yl_pred1 = model1a.predict(Xl_test)

In [105]:
print("MAE - train_test_split:", mean_absolute_error(yl_test, yl_pred1))

MAE - train_test_split: 3.111507372736363


In [19]:
from sklearn.metrics import mean_absolute_percentage_error

In [20]:
yl_pred = model1.predict(Xl)

In [21]:
print("MAPE:", mean_absolute_percentage_error(yl, yl_pred))

MAPE: 124788678399332.44


In [106]:
print("MAPE - train_test_split:", mean_absolute_percentage_error(yl_test, yl_pred1))

MAPE - train_test_split: 113680951221669.22


In [22]:
from sklearn.metrics import mean_squared_error

In [23]:
yl_pred = model1.predict(Xl)

In [24]:
print("MSE:", mean_squared_error(yl, yl_pred))

MSE: 14.722659352446817


In [107]:
print("MSE - train_test_split:", mean_squared_error(yl_test, yl_pred1))

MSE - train_test_split: 14.621726777622877


In [25]:
print("RMSE:", mean_squared_error(yl, yl_pred, squared=False))

RMSE: 3.837011773821761


In [108]:
print("RMSE - train_test_split:", mean_squared_error(yl_test, yl_pred1, squared=False))

RMSE - train_test_split: 3.823836656765411


                                       Regresja liniowa wielu zmiennych:
                                       
                                        Multivariate linear regression:

In [26]:
Xlw = dfMod[['temp', 'foggy']]
ylw = dfMod['vis']

In [27]:
Xlw.values[:5]

array([[9.47, 0.  ],
       [9.36, 0.  ],
       [9.38, 0.  ],
       [8.29, 0.  ],
       [8.76, 0.  ]])

In [28]:
ylw.values[:5]

array([15.83, 15.83, 14.96, 15.83, 15.83])

In [29]:
model2 = LinearRegression()
model2.fit(Xlw, ylw)

LinearRegression()

In [30]:
model2.intercept_

9.783870924867495

In [31]:
model2.coef_

array([ 0.10048846, -8.37535019])

In [32]:
predict_data2 = np.array([[10, 0], [12, 1]])
predict_data2

array([[10,  0],
       [12,  1]])

In [33]:
model2.predict(predict_data2)

C:\Users\Kamil - IT\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([10.78875555,  2.61438228])

In [34]:
ylw_predict = model2.predict(Xlw)

In [35]:
print("R2:", r2_score(ylw, ylw_predict))

R2: 0.4048400203594691


In [36]:
ylw_pred = model2.predict(Xlw)

In [37]:
print("MAE:", mean_absolute_error(ylw, ylw_pred))

MAE: 2.6096362816700096


In [93]:
from sklearn.model_selection import train_test_split

In [94]:
Xlw_train, Xlw_test, ylw_train, ylw_test = train_test_split(Xlw, ylw, test_size=0.2, random_state=42)

In [95]:
model2a = LinearRegression()
model2a.fit(Xlw_train, ylw_train)

LinearRegression()

In [96]:
ylw_pred1 = model2a.predict(Xlw_test)

In [97]:
print("MAE - train_test_split:", mean_absolute_error(ylw_test, ylw_pred1))

MAE - train_test_split: 2.612868902719674


In [38]:
print("MAPE:", mean_absolute_percentage_error(ylw, ylw_pred))

MAPE: 106417835106020.56


In [98]:
print("MAPE - train_test_split:", mean_absolute_percentage_error(ylw_test, ylw_pred1))

MAPE - train_test_split: 108122486316004.84


In [39]:
print("MSE:", mean_squared_error(ylw, ylw_pred))

MSE: 10.369186940914343


In [99]:
print("MSE - train_test_split:", mean_squared_error(ylw_test, ylw_pred1))

MSE - train_test_split: 10.402724254452686


In [40]:
print("RMSE:", mean_squared_error(ylw, ylw_pred, squared=False))

RMSE: 3.2201221934756363


In [100]:
print("RMSE - train_test_split:", mean_squared_error(ylw_test, ylw_pred1, squared=False))

RMSE - train_test_split: 3.225325449385331


                                               Regresja wielomianowa:
                                               
                                               Polynomial regression:

In [41]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import train_test_split

In [42]:
Xw = dfMod[['temp', 'hum', 'windsp', 'pres', 'rain', 'snow', 'mosclou', 'parclou', 'overcast', 'clear', 'foggy']]
yw = dfMod['vis']

In [43]:
Xw.values[:5]

array([[9.47000e+00, 8.90000e-01, 1.41200e+01, 1.01513e+03, 1.00000e+00,
        0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00],
       [9.36000e+00, 8.60000e-01, 1.42600e+01, 1.01563e+03, 1.00000e+00,
        0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00],
       [9.38000e+00, 8.90000e-01, 3.93000e+00, 1.01594e+03, 1.00000e+00,
        0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00],
       [8.29000e+00, 8.30000e-01, 1.41000e+01, 1.01641e+03, 1.00000e+00,
        0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00],
       [8.76000e+00, 8.30000e-01, 1.10400e+01, 1.01651e+03, 1.00000e+00,
        0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00]])

In [44]:
yw.values[:5]

array([15.83, 15.83, 14.96, 15.83, 15.83])

In [45]:
poly = PolynomialFeatures(degree=3)
Xw_poly = poly.fit_transform(Xw)

In [46]:
Xw_poly[:5]

array([[1.  , 9.47, 0.89, ..., 0.  , 0.  , 0.  ],
       [1.  , 9.36, 0.86, ..., 0.  , 0.  , 0.  ],
       [1.  , 9.38, 0.89, ..., 0.  , 0.  , 0.  ],
       [1.  , 8.29, 0.83, ..., 0.  , 0.  , 0.  ],
       [1.  , 8.76, 0.83, ..., 0.  , 0.  , 0.  ]])

In [47]:
model3 = LinearRegression()
model3.fit(Xw_poly, yw)

LinearRegression()

In [48]:
model3.intercept_

7.067057059385208

In [49]:
model3.coef_

array([-2.92364092e-07,  2.72635327e-01, -1.90037600e+01,  8.48628374e-02,
       -4.15470455e-01,  5.53542994e-01, -5.53542965e-01,  3.37171807e+00,
        2.19017510e+00,  1.73452849e+00,  1.99956877e-01,  3.44958576e+00,
       -1.90686263e-03, -1.01869626e-01,  4.91124458e-04, -6.11567257e-04,
        8.44452483e-02,  1.88198462e-01, -2.31334755e-01, -1.74634078e-01,
       -2.06625478e-01, -1.70797168e-01, -2.02981789e-01,  6.11000046e+01,
        3.37291836e-02,  5.28161053e-02, -1.63186252e+01, -2.68512793e+00,
       -1.10353654e+01, -6.96044200e+00, -5.02408592e+00, -2.48994176e-01,
       -1.99389653e+01, -2.02762323e-03, -6.17875360e-04,  1.79371867e-02,
        6.69257446e-02, -7.45823786e-02, -2.44542324e-02, -5.46336772e-02,
       -1.09054666e-02, -1.87913998e-01,  1.10955657e-03, -2.01411141e-01,
       -2.14059486e-01, -1.74533883e-02, -2.01490294e-02, -1.85389726e-02,
       -1.68003046e-02, -3.54921144e-02,  5.53542921e-01, -6.50889342e-11,
        3.51334718e+00,  

In [50]:
predict_data3 = np.array([[15, 0.65, 19, 1012, 0, 0, 0, 1, 0, 0, 0], [16.5, 0.33, 22.5, 1003, 0, 0, 0, 0, 0, 1, 0]])
predict_data3

array([[1.500e+01, 6.500e-01, 1.900e+01, 1.012e+03, 0.000e+00, 0.000e+00,
        0.000e+00, 1.000e+00, 0.000e+00, 0.000e+00, 0.000e+00],
       [1.650e+01, 3.300e-01, 2.250e+01, 1.003e+03, 0.000e+00, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00, 1.000e+00, 0.000e+00]])

In [51]:
predict_data_poly = poly.transform(predict_data3)

C:\Users\Kamil - IT\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but PolynomialFeatures was fitted with feature names
  warnings.warn(


In [52]:
predict_data_poly

array([[1.00000000e+00, 1.50000000e+01, 6.50000000e-01, 1.90000000e+01,
        1.01200000e+03, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        1.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        2.25000000e+02, 9.75000000e+00, 2.85000000e+02, 1.51800000e+04,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 1.50000000e+01,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 4.22500000e-01,
        1.23500000e+01, 6.57800000e+02, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00, 6.50000000e-01, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00, 3.61000000e+02, 1.92280000e+04, 0.00000000e+00,
        0.00000000e+00, 0.00000000e+00, 1.90000000e+01, 0.00000000e+00,
        0.00000000e+00, 0.00000000e+00, 1.02414400e+06, 0.00000000e+00,
        0.00000000e+00, 0.00000000e+00, 1.01200000e+03, 0.00000000e+00,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.000000

In [53]:
model3.predict(predict_data_poly)

array([-138.4250869 , -122.64568908])

In [ ]:
# Występują wartości irracjonalne. 

# Irrational values occur. 

In [54]:
yw_predict = model3.predict(Xw_poly)

In [55]:
print("R2:", r2_score(yw, yw_predict))

R2: 0.5454334429263288


In [56]:
print("MAE:", mean_absolute_error(yw, yw_predict))

MAE: 2.330420842347345


In [84]:
from sklearn.model_selection import train_test_split

In [86]:
Xw_poly_train, Xw_poly_test, yw_train, yw_test = train_test_split(Xw_poly, yw, test_size=0.2, random_state=42)

In [87]:
model3a = LinearRegression()
model3a.fit(Xw_poly_train, yw_train)

LinearRegression()

In [88]:
yw_predict1 = model3a.predict(Xw_poly_test)

In [89]:
print("MAE - train_test_split:", mean_absolute_error(yw_test, yw_predict1))

MAE - train_test_split: 2.340545761813752


In [57]:
print("MAPE:", mean_absolute_percentage_error(yw, yw_predict))

MAPE: 81083402916230.56


In [90]:
print("MAPE - train_test_split:", mean_absolute_percentage_error(yw_test, yw_predict1))

MAPE - train_test_split: 87671348721064.66


In [58]:
print("MSE:", mean_squared_error(yw, yw_predict))

MSE: 7.919695155295207


In [91]:
print("MSE - train_test_split:", mean_squared_error(yw_test, yw_predict1))

MSE - train_test_split: 8.0115106534551


In [59]:
print("RMSE:", mean_squared_error(yw, yw_predict, squared=False))

RMSE: 2.814195294448345


In [92]:
print("RMSE - train_test_split:", mean_squared_error(yw_test, yw_predict1, squared=False))

RMSE - train_test_split: 2.8304612086116108


                                             Drzewo decyzyjne (wielu zmiennych):
                                             
                                               Decision Tree (multi-variable):

In [60]:
from sklearn.tree import DecisionTreeRegressor

In [61]:
Xd = dfMod[['temp', 'hum', 'windsp', 'pres', 'rain', 'snow', 'mosclou', 'parclou', 'overcast', 'clear', 'foggy']]
yd = dfMod['vis']

In [62]:
Xd.values[:5]

array([[9.47000e+00, 8.90000e-01, 1.41200e+01, 1.01513e+03, 1.00000e+00,
        0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00],
       [9.36000e+00, 8.60000e-01, 1.42600e+01, 1.01563e+03, 1.00000e+00,
        0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00],
       [9.38000e+00, 8.90000e-01, 3.93000e+00, 1.01594e+03, 1.00000e+00,
        0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00],
       [8.29000e+00, 8.30000e-01, 1.41000e+01, 1.01641e+03, 1.00000e+00,
        0.00000e+00, 0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00],
       [8.76000e+00, 8.30000e-01, 1.10400e+01, 1.01651e+03, 1.00000e+00,
        0.00000e+00, 1.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
        0.00000e+00]])

In [63]:
yd.values[:5]

array([15.83, 15.83, 14.96, 15.83, 15.83])

In [64]:
model4 = DecisionTreeRegressor(max_depth=5, min_samples_split=3)
model4.fit(Xd, yd)

# max_depth - ile poziomów drzewo może maksymalnie zawierać.
#           - how many levels the tree can contain a most. 

# min_samples_split - ile minimalnie przykłądów musi być w węźle tak, by dokonać kolejnego podziału.
#                   - how many minimum examples must be in a node to make another division. 

DecisionTreeRegressor(max_depth=5, min_samples_split=3)

In [65]:
predict_data4 = np.array([[15, 0.65, 19, 1012, 0, 0, 0, 1, 0, 0, 0], [16.5, 0.33, 22.5, 1003, 0, 0, 0, 0, 0, 1, 0]])
predict_data4

array([[1.500e+01, 6.500e-01, 1.900e+01, 1.012e+03, 0.000e+00, 0.000e+00,
        0.000e+00, 1.000e+00, 0.000e+00, 0.000e+00, 0.000e+00],
       [1.650e+01, 3.300e-01, 2.250e+01, 1.003e+03, 0.000e+00, 0.000e+00,
        0.000e+00, 0.000e+00, 0.000e+00, 1.000e+00, 0.000e+00]])

In [66]:
model4.predict(predict_data4)

C:\Users\Kamil - IT\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:439: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(


array([11.92258962, 11.92258962])

In [67]:
yd_predict = model4.predict(Xd)

In [68]:
print("R2:", r2_score(yd, yd_predict))

R2: 0.5274905182601561


In [69]:
print("MAE:", mean_absolute_error(yd, yd_predict))

MAE: 2.3438002256161004


In [73]:
from sklearn.model_selection import train_test_split

In [74]:
Xd_train, Xd_test, yd_train, yd_test = train_test_split(Xd, yd, test_size=0.2, random_state=42)

In [75]:
model4a = DecisionTreeRegressor(max_depth=5, min_samples_split=3)
model4a.fit(Xd_train, yd_train)

DecisionTreeRegressor(max_depth=5, min_samples_split=3)

In [76]:
yd_pred = model4a.predict(Xd_test)

In [79]:
print("MAE - train_test_split:", mean_absolute_error(yd_test, yd_pred))

MAE - train_test_split: 2.3523298038998126


In [70]:
print("MAPE:", mean_absolute_percentage_error(yd, yd_predict))

MAPE: 94605787205310.33


In [80]:
print("MAPE - train_test_split:", mean_absolute_percentage_error(yd_test, yd_pred))

MAPE - train_test_split: 100569516978913.45


In [71]:
print("MSE:", mean_squared_error(yd, yd_predict))

MSE: 8.23230612796622


In [81]:
print("MSE - train_test_split:", mean_squared_error(yd_test, yd_pred))

MSE - train_test_split: 8.290791782002515


In [72]:
print("RMSE:", mean_squared_error(yd, yd_predict, squared=False))

RMSE: 2.869199562241396


In [82]:
print("RMSE - train_test_split:", mean_squared_error(yd_test, yd_pred, squared=False))

RMSE - train_test_split: 2.8793735051226883


                                                      Podsumowanie:
                                                      
                                                        Summary:

  Zarówno Drzewo Decyzyjne, jak i regresja wielomianowa osiągają podobny poziom wyników (średni - 53 - 55 %). Mamy więc do czynienia z tzw. niedotrenowaniem, a sposobami na jego rozwiązanie mogą być:
  
  - wybór innego modelu (np. sieć neuronowa rekurencyjna),
  - wyłączenie pewnych cech (mniej istotnych),
  - optymalizacja hiperparametrów,
  - inna kompozycja danych wejściowych.
  
  
  Both Decision Tree and Polynomial Regression achieve similar levels of performance (average 53 - 55 %). So we are  dealing 
with the so-called undertraining, and ways to solve it may be:

  - choosing another model (e.g. recurrent neural network),
  - exclusion of certain features (less important),
  - hyperparameter optimization,
  - different composition of input data.
  